### ISSUE TITLE
Uneven VRAM usage with sharded model serving

### ISSUE THREAD

**Issue by ipoletaev** [huggingface/text-generation-inference#776]
### System Info

Running 0.9.4 container on ubuntu `20.04.6 LTS` with CUDA 11.8 and A100s.

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Reproduction

```
docker run --gpus=all --rm --shm-size 1g \
    -p $port:$port ghcr.io/huggingface/text-generation-inference:0.9.4 \
    --model-id llama-70b --num-shard 4 \
    --max-input-length 4095 \
    --max-total-tokens 4096\
    --max-batch-prefill-tokens 32768 \
    --cuda-memory-fraction 0.99 \
    --hostname 0.0.0.0 --port $port
```

### Expected behavior

While serving default fp16 LLaMA-2-70B using 4 or 8 shards and logging from each shards information about free memory available I see (below is an example of 4 shards):

```
2023-08-06T05:42:44.436040Z  INFO text_generation_launcher: Rank [1] Total GPU memory: 79.21 GB, 
Total free memory: 5.32 GB, 
Effective supported batch size: 22

2023-08-06T05:42:44.453502Z  INFO text_generation_launcher: Rank [0] Total GPU memory: 79.21 GB, 
Total free memory: 5.46 GB, 
Effective supported batch size: 22

2023-08-06T05:42:44.468010Z  INFO text_generation_launcher: Rank [3] Total GPU memory: 79.21 GB, 
Total free memory: 2.71 GB, 
Effective supported batch size: 14

2023-08-06T05:42:44.479387Z  INFO text_generation_launcher: Rank [2] Total GPU memory: 79.21 GB, 
Total free memory: 5.32 GB, 
Effective supported batch size: 22
```

The above is printed [over here](https://github.com/huggingface/text-generation-inference/blob/16fadcec5711ff232977b38c74a1c8829af6a63b/server/text_generation_server/models/flash_causal_lm.py#L768) as follows:

```
        max_supported_batch_total_tokens = int(num_blocks * BLOCK_SIZE)
        logger.info(
            f"Rank [{self.rank}] Total GPU memory: {total_gpu_memory / 1024 ** 3:.2f} GB, \n"
            f"Total free memory: {total_free_memory / 1024 ** 3:.2f} GB, \n"
            f"Total free memory allowed: {free_memory / 1024 ** 3:.2f} GB, \n"
            f"Effective supported batch size: {int(max_supported_batch_total_tokens/max_seqlen)}"
        )
```        

In other words each shard sees different VRAM available and thereby returns back to the router different value for `max_supported_batch_total_tokens`. Hence it is unclear whether (1) it is expected (2) even if not - is router OK to deal with such imbalanced capabilities of the shards?

Thanks!

**Comment by Narsil** [comment]
Is this consistent across runs ? Maybe 1 shard got some fragmentation going on and that's what we're seeing ?

**Comment by Narsil** [comment]
I'm unable to reproduce, is there any way there's already a running process on GPU3 ?

Can you provide the output of `nvidia-smi` after the load ?

**Comment by ipoletaev** [comment]
Sure. Here is an example with the same serving h-params:

```
2023-08-09T16:39:17.557727Z  INFO text_generation_router: router/src/main.rs:210: Warming up model
2023-08-09T16:39:30.032821Z  INFO text_generation_launcher: Rank [0] Total GPU memory: 79.21 GB, 
Total free memory: 7.84 GB, 
Total free memory allowed: 7.05 GB, 
Effective supported batch size: 30

2023-08-09T16:39:30.051492Z  INFO text_generation_launcher: Rank [2] Total GPU memory: 79.21 GB, 
Total free memory: 7.70 GB, 
Total free memory allowed: 6.91 GB, 
Effective supported batch size: 30

2023-08-09T16:39:30.076165Z  INFO text_generation_launcher: Rank [3] Total GPU memory: 79.21 GB, 
Total free memory: 5.09 GB, 
Total free memory allowed: 4.30 GB, 
Effective supported batch size: 21

2023-08-09T16:39:30.078729Z  INFO text_generation_launcher: Rank [1] Total GPU memory: 79.21 GB, 
Total free memory: 2.57 GB, 
Total free memory allowed: 1.78 GB, 
Effective supported batch size: 13
```

Idling:

```
|===============================+======================+======================|
|   0  NVIDIA A100-SXM...  On   | 00000000:0F:00.0 Off |                    0 |
| N/A   37C    P0    83W / 400W |  45202MiB / 81920MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM...  On   | 00000000:15:00.0 Off |                    0 |
| N/A   34C    P0    81W / 400W |  39912MiB / 81920MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   2  NVIDIA A100-SXM...  On   | 00000000:50:00.0 Off |                    0 |
| N/A   35C    P0    83W / 400W |  45346MiB / 81920MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   3  NVIDIA A100-SXM...  On   | 00000000:53:00.0 Off |                    0 |
| N/A   35C    P0    81W / 400W |  42340MiB / 81920MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
```

Under peak load:

```
|===============================+======================+======================|
|   0  NVIDIA A100-SXM...  On   | 00000000:0F:00.0 Off |                    0 |
| N/A   65C    P0   416W / 400W |  80834MiB / 81920MiB |     81%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM...  On   | 00000000:15:00.0 Off |                    0 |
| N/A   57C    P0   426W / 400W |  80814MiB / 81920MiB |    100%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   2  NVIDIA A100-SXM...  On   | 00000000:50:00.0 Off |                    0 |
| N/A   58C    P0   397W / 400W |  80978MiB / 81920MiB |    100%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   3  NVIDIA A100-SXM...  On   | 00000000:53:00.0 Off |                    0 |
| N/A   62C    P0   377W / 400W |  80758MiB / 81920MiB |    100%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
```

The other scary thing I found is that regular RAM leaks very slowly... ~1GB/day with ~1QPS sent to such a 4-shards server.

**Comment by ipoletaev** [comment]
I see, thanks.
Wonder if @Narsil you have had some meaningful time spent on figuring out if it is debuggable at all? 
I don't want to think there is nothing one can do but just to live with that :)

**Comment by Narsil** [comment]
If I can't reproduce, it's really hard to debug you know.

I have it in mind, so I'll keep looking for it.

**Comment by Narsil** [comment]
I'm going to close this, since the uneven part of VRAM usage doesn't seem to be that bad.

Please do comment if you see it cropping up again, or have an easier way to reproduce so we can actually fix.
Our cache manager would be the place to look if anyone's interested: https://github.com/huggingface/text-generation-inference/blob/0e8b47811e711b46fe80e6b4f5304186f83744d6/server/text_generation_server/models/flash_causal_lm.py#L31
